### ISSUE TITLE
false error on generic type alias parameters

### ISSUE THREAD

**Issue by Eliav2** [microsoft/TypeScript#58253]
### 🔎 Search Terms

false error on generic type alias parameters

### 🕗 Version & Regression Information

ts v5.4.5

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/KYDwDg9gTgLgBDAnmYcAqzgBMBKECuMwUAPAFBxw5yhEB2WAznAN4WVwDaA1sIgFxxGMKAEs6AcwC6ggBQBDKBMaD5dRAEo4AXgB8cNYgDc7AL5l92qibKhIsBJjgBxYDDyFgOYI0h1GwACSdABmEOSUHkRQNCD0TOiYuATRJIa6ADTsAAryMAAWsfHMvIgQIVQpxBY6VG74UHQYKCRRxJy5BVK6JgD0vRyDQ3AAeokocADkbVCcwmKSUpNwosx0EPDyjIyiEnTyAEYANqgwEI4Tk7IAdLeKyqrqWnoG6pNktuDQ8EgTzdgzABiohA2AiVCKwAYzDYgx4fEE83E0jk9xUr00On0hhMlHMllYUCqKhwphsdm+F1QrncVW8vgg-iCoQgwNBWHBM0h0PGAKqUDZYPSWUonUKtChCVK5UqnlmkyJnkYkykNSs3hgDSamFa-M4CuJKo6eXy3T6A2GwzGAHdoNxGABCD4Uhy-aluNr0vwBYJhABqxAOEACnP53ISsI48IEQhEyJkcAUSnRhme2PUuLgphFcDF4ZKfBlMzVdU1jX+urlxq6PTI-UtVoMR0Y51tUHtGTgB0IcAAtvgAMaFXvQVAAN0DweA1yAA

### 💻 Code

```ts
export type TypedRouter<
  R extends {
    [key: string]: (args: any) => any;
  }
> = R;

export type GetRouteResponseInfo<
  Router extends TypedRouter<any>,
  Path extends keyof Router
> = ReturnType<Router[Path]>;
//                        ^ Type 'Router[string]' is not assignable to type '(...args: any) => any'
```

### 🙁 Actual behavior

in `GetRouteResponseInfo`, `Router[Path]` is not considered a function even though `Router` is `TypedRouter` which accepts only parameters that satisfy this constrain.

ironically, returning the type inside nested prop relaxes the error:
```ts
export type TypedRouterFixed<
  R extends {
    [key: string]: (args: any) => any;
  }
> = {routes:R};

export type GetRouteResponseInfoFixed<
  Router extends TypedRouterFixed<any>,
  Path extends keyof Router['routes']
> = ReturnType<Router['routes'][Path]>;
//                         ^ no error!
``` 

i know that specifying the constraint directly instead of using other generic such `TypedRouter` relaxes the issue also:
```ts
export type TypedRouter<
  R extends {
    [key: string]: (args: any) => any;
  }
> = R;

export type GetRouteResponseInfo<
  Router extends {
    [key: string]: (args: any) => any;
  },
  Path extends keyof Router
> = ReturnType<Router[Path]>;
//                         ^ no error!
```

however, that should not be necessary as any T in `TypedRouter<T>` would already satisfy `{[key: string]: (args:any) => any}`, and the example with the nested `routes` proves it.

### 🙂 Expected behavior

no error should be reported.

### Additional information about the issue

the real use case is with more complected type, and currently we are hiding this prop inside nested prop and extract this type from the nested from to work around this issue.

**Comment by RyanCavanaugh** [comment]
This is a correct error; you could write
```ts
type MyRouter = {
  [Symbol.isConcatSpreadable]: number;
}
type M = GetRouteResponseInfo<MyRouter, typeof Symbol.isConcatSpreadable>;
```
at which point `Router[Path]` has type `number`, which there's no `ReturnType<>` of

**Comment by Eliav2** [comment]
hey @RyanCavanaugh, I'm not sure you understand my point.

`TypedRouter` retruns R which is contained to be at least `{[key: string]: (args: any) => any;}` (even if any is passed, like in `TypedRouter<any>`), and because `Router extends TypedRouter<any>`, then Router is also constrained to `{[key: string]: (args: any) => any;}`. so `keyof Router` would forever be `string`, and `Router[Path]` would forever be a valid function.

In your example, You just assumed a route key of type `number` which is not possible (because the constrain on R in `TypedRouter`).  your example that not explain why this is 'Working as Intended` at all. 

and if so, why does holding the type inside nested prop fixes the issue and the compiler does not complain anymore?

**Comment by RyanCavanaugh** [comment]
>  then Router is also constrained to `{[key: string]: (args: any) => any;}`

This index signature only constrains what's present at string keys. It doesn't bar what happens at symbol keys
```ts
type MyRouter = {
  [Symbol.isConcatSpreadable]: number;
}
type IsRouter<T extends TypedRouter<any>> = T;
type Test = IsRouter<MyRouter>; // Not a constraint violation
```

**Comment by Eliav2** [comment]
I think I get it now, and it does not relates to the way index signature checks symbols.
actually, in your example you can use `'random'` literal (or any other random type), and it wouldn't be a constrain violation.
```ts
export type TypedRouter<
  R extends {
    [key: string]: (args: any) => any;
  }
> = R;

type IsRouter<T extends TypedRouter<any>> = T;

type Test = IsRouter<'random'>; // <-- Not a constraint violation
```
this is because when checking for `T extends TypedRouter<any>`, typescript checks if R=`any` extends `{[key: string]: (args: any) => any;}` (and it does), so is it returns R(`any`) without farther constraining it to `{[key: string]: (args: any) => any;}`, which we can debate if is excepted(for user perspective) or not.

---

in other words,
```ts
type IsRouter<T extends TypedRouter<any>> = T;
```
is resolved to
```ts
type IsRouter<T extends any> = T;
```
(because `TypedRouter<any>=any`)
instead of 
```ts
type IsRouter<T extends {[key: string]: (args: any) => any}> = T;
```
(because `any` in `TypedRouter<any>` should be narrowed down to `{[key: string]: (args: any) => any}` as `R` in `TypedRouter` applies)

i would claim that this is not expected, and the type should be narrowed down(but that just my opinion 🫠). 

nevertheless, because the type are not narrowed down, you are correct that this works as intended, and this issue can be closed.  

but one last thing, is there is documentation/issue that explains the design choice to not narrowing down returned generic types by their constraint? (in other words, why `TypedRouter<any>=any` instead of `{[key: string]: (args: any) => any}`)
